lib/vssolution.cpp

```diff
@@ -46,77 +46,77 @@ int isfinite(double x)
-        << "| o -  (De)refine elem. (NC shading) |" << endl
-        << "| O -  Switch 'o' func. (NC shading) |" << endl
-        << "| p/P  Cycle through color palettes  |" << endl
-        << "| q -  Quits                         |" << endl
-        << "| r -  Reset the plot to 3D view     |" << endl
-        << "| R -  Reset the plot to 2D view     |" << endl
-        << "| s -  Turn on/off unit cube scaling |" << endl
-        << "| S -  Take snapshot/Record a movie  |" << endl
-        << "| t -  Cycle materials and lights    |" << endl
-        << "| y/Y  Rotate the cutting plane      |" << endl
-        << "| z/Z  Move the cutting plane        |" << endl
-        << "| \\ -  Set light source position     |" << endl
-        << "| Ctrl+o - Element ordering curve    |" << endl
-        << "| Ctrl+p - Print to a PDF file       |" << endl
-        << "+------------------------------------+" << endl
-        << "| Function keys                      |" << endl
-        << "+------------------------------------+" << endl
-        << "| F1 - X window info and keystrokes  |" << endl
-        << "| F2 - Update colors, etc.           |" << endl
-        << "| F3/F4 - Shrink/Zoom elements       |" << endl
-        << "| F5 - Set level lines               |" << endl
-        << "| F6 - Palette options               |" << endl
-        << "| F7 - Manually set min/max value    |" << endl
-        << "| F8 - List of subdomains to show    |" << endl
-        << "| F9/F10 - Walk through subdomains   |" << endl
-        << "| F11/F12 - Shrink/Zoom subdomains   |" << endl
-        << "+------------------------------------+" << endl
-        << "| Keypad                             |" << endl
-        << "+------------------------------------+" << endl
-        << "| 1-9  Small rotation, reset with 5  |" << endl
-        << "| *,/  Scale up/down                 |" << endl
-        << "| +/-  Change z-scaling              |" << endl
-        << "| . -  Start/stop spinning           |" << endl
-        << "| 0/Enter - Spinning speed and dir.  |" << endl
-        << "+------------------------------------+" << endl
-        << "| Mouse                              |" << endl
-        << "+------------------------------------+" << endl
-        << "| left   btn    - Rotation           |" << endl
-        << "| middle btn    - Translation        |" << endl
-        << "| right  btn    - Scaling            |" << endl
-        << "| left  + Alt   - Tilt               |" << endl
-        << "| left  + Shift - Spinning           |" << endl
-        << "| right + Shift - Change light pos.  |" << endl
-        << "| left  + Ctrl  - Spherical rotation |" << endl
-        << "| middle+ Ctrl  - Object translation |" << endl
-        << "| right + Ctrl  - Object scaling     |" << endl
-        << "| left  + Ctrl + Shift - z-Spinning  |" << endl
-        << "+------------------------------------+" << endl;
+static void SolutionHelpFunction(std::ostream & os)
```

Comment:
We could go one step further, and make these member virtual functions on the VisualizationSceneSolution classes that return an `std::string`. Then we could avoid having to go through the SdlWindow to get the help string.

Reply:
I agree.

I don't remember anymore why these functions were declared to be static. Maybe there was a reason for it with the old X11 code, but if it is not necessary anymore having them as member functions will be much cleaner.

@kanye-quest do you want to make the change? (I can make it too)

Reply:
I think the issue with this in the current code would be that `SdlWindow::onKeyDown` uses function pointers so a method wouldn't work, but it could be changed to use `std::function` so we could use a lambda with the captured `VisualizationSceneSolution` instead